xof.go

```diff
@@ -0,0 +1,53 @@
+package kyber
+
+import (
+	"crypto/cipher"
+)
+
+// Sponge is an interface representing a primitive sponge function.
+type Sponge interface {
+	// XOR src data into sponge's internal state,
+	// transform its state, and copy resulting state into dst.
+	// Buffers must be either Rate or Rate+Capacity bytes long.
+	Transform(dst, src []byte)
+
+	// Return the number of data bytes the sponge can aborb in one block.
+	Rate() int
+
+	// Return the sponge's secret state capacity in bytes.
+	Capacity() int
+
+	// Clone clones a Sponge.
+	Clone() Sponge
+}
+
+// Xof is an interface for extendable output functions.
+type Xof interface {
+	// Absorb introduces state into the internal state of the Xof.
+	// Use it to put in keying material, or to put in material to be
+	// hashed. If key does not fill the internal state until Rate(),
+	// it will be padded with zeros until it does. Absorb returns
+	// itself so that chaining during creation is possible:
+	//     x := xof.New("keccak").Absorb(seed)
+	Absorb(key []byte) Xof
```

Comment:
Perhaps the "absorb" method(s) should be made compatible with the corresponding ones in the Go standard Hash interface?

Reply:
In fact, a fairly reasonable potential definition of this interface might be:

type XOF interface {
    hash.Hash
    cipher.Stream
}

...maybe along with a Clone method added.  That way an XOF can be used anywhere a hash is used, which makes sense; in that usage the Sum() and Size() methods should return the "default" output sizes of the XOF, which should be the minimum sizes to provide strong security (namely 2x the bits of the base security parameter, e.g., 256 bits for SHAKE128 or 512 bits for SHAKE256).  So anyone can use the XOF as a Hash if they want to, but can alternatively Write (absorb) stuff into it on input and then start calling XORKeyStream (using it as a cipher.Stream) to switch into output mode and suck an arbitrary amount of data out of it incrementally.  Does that make sense?